pkg/querier/vcs/client/github.go

```diff
@@ -15,21 +15,26 @@ import (
 )
 
 // GithubClient returns a github client.
-func GithubClient(ctx context.Context, token *oauth2.Token) (*githubClient, error) {
+func GithubClient(ctx context.Context, token *oauth2.Token, metrics Metrics) (*githubClient, error) {
 	return &githubClient{
-		client: github.NewClient(nil).WithAuthToken(token.AccessToken),
+		client:  github.NewClient(nil).WithAuthToken(token.AccessToken),
```

Comment:
My feeling is that a more straightforward way of getting the actual github status code instrumented would be by using an instrumented `http.Client` (for NewClient's parameter) with a middleware that could collect the same metrics.

The prometheus client already comes with ones e.g. this one for duration:

https://github.com/prometheus/client_golang/blob/77d4003c72f054ac435df1223deac17b1f8858ea/prometheus/promhttp/instrument_client.go#L87

We do something for tracing already here (we don't use it for the github): https://github.com/grafana/pyroscope/blob/09f0787e8a0f049b36ffa105161aa66bdb3c9e89/pkg/util/http.go#L53

Reply:
I did think about this. I agree with you that the current approach is very brittle and doesn't provide very good access to other information surrounding the request. I think adding `RoundTripper` middleware is an excellent solution.

The main drawback to that approach is I'm not sure how to label the metrics with the API path template. Using middleware we can get the complete path, but that includes ids, shas/ filenames, etc. E.g. I want

```
/repos/{owner}/{repo}/commits/{ref}
```

but the middleware would provide

```
/repos/grafana/pyroscope/commits/58f66fc6e6806101fa3eb9cc01f73a8c50b8aed5
```

Reply:
I think something to adopt from dskit: https://github.com/grafana/dskit/blob/c41c58734f46e8d4b5f1f407a6a8c01eb4184d17/middleware/instrument.go#L131